File: scripts/benchmark_failure_dashboard.py
#!/usr/bin/env python3
"""
Generate nightly benchmark failure-class dashboard artifacts.

Outputs:
- <output-dir>/latest_failure_dashboard.json
- <output-dir>/latest_failure_dashboard.md
"""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
from typing import Any, Dict, Iterable, List, Tuple


SUCCESS_REASONS = {
    "none",
    "completed",
    "critical_findings_detected",
}

FAILURE_CLASS_RULES: List[Tuple[str, set[str]]] = [
    ("lock_contention", {"output_dir_locked"}),
    (
        "setup_tooling",
        {
            "backend_tooling_missing",
            "backend_preflight_failed",
            "circom_compilation_failed",
            "key_generation_failed",
            "missing_invariants",
            "readiness_failed",
            "filesystem_permission_denied",
        },
    ),
    ("timeouts", {"wall_clock_timeout"}),
    (
        "stability_runtime",
        {
            "runtime_error",
            "panic",
            "artifact_mirror_panic_missing_command",
            "run_outcome_missing",
            "run_outcome_unreadable",
            "run_outcome_invalid_json",
            "unknown",
            "stale_interrupted",
        },
    ),
    (
        "contract_or_config",
        {
            "engagement_contract_failed",
            "missing_chains_definition",
        },
    ),
]

CLASS_THRESHOLDS = {
    "lock_contention": 0.05,
    "setup_tooling": 0.15,
    "timeouts": 0.10,
    "stability_runtime": 0.05,
    "contract_or_config": 0.10,
    "other_failure": 0.05,
}

FAILURE_CLASSES = [
    "lock_contention",
    "setup_tooling",
    "timeouts",
    "stability_runtime",
    "contract_or_config",
    "other_failure",
]


def _latest_benchmark_dir(benchmark_root: Path) -> Path:
    candidates = sorted(p for p in benchmark_root.glob("benchmark_*") if p.is_dir())
    if not candidates:
        raise FileNotFoundError(f"No benchmark_* directory found under {benchmark_root}")
    return candidates[-1]


def _load_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def _classify_reason(reason: str) -> str:
    if reason in SUCCESS_REASONS:
        return "success"
    for class_name, reasons in FAILURE_CLASS_RULES:
        if reason in reasons:
            return class_name
    return "other_failure"


def _sum_reason_counts(outcomes: Iterable[Dict[str, Any]]) -> Dict[str, int]:
    counts: Dict[str, int] = {}
    for item in outcomes:
        reason_counts = item.get("reason_counts", {})
        if not isinstance(reason_counts, dict):
            continue
        for reason, raw_count in reason_counts.items():
            try:
                count = int(raw_count)
            except (TypeError, ValueError):
                continue
            counts[reason] = counts.get(reason, 0) + max(count, 0)
    return counts


def _pct(value: float) -> str:
    return f"{value * 100.0:.2f}%"


def _parse_rate(raw: str, source: str) -> float:
    try:
        value = float(raw)
    except ValueError as exc:
        raise ValueError(f"Invalid threshold value from {source}: {raw!r}") from exc
    if value < 0.0 or value > 1.0:
        raise ValueError(f"Threshold from {source} must be between 0.0 and 1.0, got {value}")
    return value


def _env_key_for_class(class_name: str) -> str:
    return f"ZKF_FAILURE_MAX_RATE_{class_name.upper()}"


def _resolve_thresholds(cli_overrides: List[str]) -> Dict[str, float]:
    thresholds = dict(CLASS_THRESHOLDS)

    for class_name in FAILURE_CLASSES:
        env_key = _env_key_for_class(class_name)
        raw_env = os.environ.get(env_key)
        if raw_env is None or raw_env.strip() == "":
            continue
        thresholds[class_name] = _parse_rate(raw_env.strip(), f"${env_key}")

    for override in cli_overrides:
        if "=" not in override:
            raise ValueError(f"Invalid --threshold {override!r}; expected CLASS=RATE")
        class_name_raw, rate_raw = override.split("=", 1)
        class_name = class_name_raw.strip().lower()
        if class_name not in thresholds:
            valid = ", ".join(FAILURE_CLASSES)
            raise ValueError(f"Unknown failure class {class_name!r}; valid classes: {valid}")
        thresholds[class_name] = _parse_rate(
            rate_raw.strip(), f"--threshold {class_name_raw.strip()}={rate_raw.strip()}"
        )

    return thresholds


def _dashboard(
    summary: Dict[str, Any],
    outcomes: List[Dict[str, Any]],
    summary_path: Path,
    outcomes_path: Path,
    class_thresholds: Dict[str, float],
) -> Dict[str, Any]:
    total_runs = int(summary.get("total_runs", len(outcomes)))
    if total_runs <= 0:
        total_runs = len(outcomes)

    reason_counts = _sum_reason_counts(outcomes)
    class_counts: Dict[str, int] = {}
    for reason, count in reason_counts.items():
        cls = _classify_reason(reason)
        class_counts[cls] = class_counts.get(cls, 0) + count

    class_rows: List[Dict[str, Any]] = []
    for class_name in FAILURE_CLASSES:
        count = class_counts.get(class_name, 0)
        rate = (count / total_runs) if total_runs > 0 else 0.0
        threshold = class_thresholds[class_name]
        status = "PASS" if rate <= threshold else "FAIL"
        class_rows.append(
            {
                "class": class_name,
                "count": count,
                "rate": rate,
                "max_rate": threshold,
                "status": status,
            }
        )

    overall_status = "PASS"
    for row in class_rows:
        if row["status"] == "FAIL":
            overall_status = "FAIL"
            break

    return {
        "generated_utc": summary.get("generated_utc"),
        "summary_path": str(summary_path),
        "outcomes_path": str(outcomes_path),
        "total_runs": total_runs,
        "overall_status": overall_status,
        "class_rows": class_rows,
        "reason_counts": dict(sorted(reason_counts.items())),
    }


def _write_markdown(path: Path, payload: Dict[str, Any]) -> None:
    lines: List[str] = []
    lines.append("# Benchmark Failure-Class Dashboard")
    lines.append("")
    lines.append(f"- Generated UTC: `{payload.get('generated_utc')}`")
    lines.append(f"- Summary: `{payload.get('summary_path')}`")
    lines.append(f"- Outcomes: `{payload.get('outcomes_path')}`")
    lines.append(f"- Overall status: `{payload.get('overall_status')}`")
    lines.append(f"- Total runs: `{payload.get('total_runs')}`")
    lines.append("")
    lines.append("| Failure Class | Count | Rate | Max Allowed | Status |")
    lines.append("|---|---:|---:|---:|---|")
    for row in payload.get("class_rows", []):
        lines.append(
            f"| {row['class']} | {row['count']} | {_pct(row['rate'])} | {_pct(row['max_rate'])} | {row['status']} |"
        )
    lines.append("")
    lines.append("## Reason Counts")
    lines.append("")
    lines.append("| Reason Code | Count |")
    lines.append("|---|---:|")
    reason_counts: Dict[str, int] = payload.get("reason_counts", {})
    for reason, count in reason_counts.items():
        lines.append(f"| {reason} | {count} |")
    lines.append("")
    path.write_text("\n".join(lines), encoding="utf-8")


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate benchmark failure-class dashboard.")
    parser.add_argument(
        "--benchmark-root",
        default="artifacts/benchmark_runs",
        help="Root directory containing benchmark_<timestamp>/summary.json and outcomes.json",
    )
    parser.add_argument(
        "--output-dir",
        default="artifacts/benchmark_trends",
        help="Directory where dashboard artifacts are written",
    )
    parser.add_argument(
        "--threshold",
        action="append",
        default=[],
        metavar="CLASS=RATE",
        help=(
            "Override class threshold (repeatable). "
            "Also supports env ZKF_FAILURE_MAX_RATE_<CLASS>, e.g. "
            "ZKF_FAILURE_MAX_RATE_SETUP_TOOLING=0.20"
        ),
    )
    args = parser.parse_args()

    benchmark_root = Path(args.benchmark_root)
    output_dir = Path(args.output_dir)
    class_thresholds = _resolve_thresholds(args.threshold)
    latest_dir = _latest_benchmark_dir(benchmark_root)
    summary_path = latest_dir / "summary.json"
    outcomes_path = latest_dir / "outcomes.json"

    if not summary_path.exists():
        raise FileNotFoundError(f"Missing summary.json at {summary_path}")
    if not outcomes_path.exists():
        raise FileNotFoundError(f"Missing outcomes.json at {outcomes_path}")

    summary = _load_json(summary_path)
    outcomes = _load_json(outcomes_path)
    if not isinstance(outcomes, list):
        raise ValueError(f"Expected list in {outcomes_path}")

    payload = _dashboard(summary, outcomes, summary_path, outcomes_path, class_thresholds)
    output_dir.mkdir(parents=True, exist_ok=True)

    json_path = output_dir / "latest_failure_dashboard.json"
    md_path = output_dir / "latest_failure_dashboard.md"
    json_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    _write_markdown(md_path, payload)

    print(f"Failure dashboard written: {json_path}")
    print(f"Failure dashboard report written: {md_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/checks/benchmark_failure_dashboard.rs
use anyhow::{bail, Context};
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};
use std::env;
use std::path::{Path, PathBuf};

pub const SUCCESS_REASONS: &[&str] = &["none", "completed", "critical_findings_detected"];

pub const FAILURE_CLASSES: &[&str] = &[
    "lock_contention",
    "setup_tooling",
    "timeouts",
    "stability_runtime",
    "contract_or_config",
    "other_failure",
];

pub const FAILURE_CLASS_RULES: &[(&str, &[&str])] = &[
    ("lock_contention", &["output_dir_locked"]),
    (
        "setup_tooling",
        &[
            "backend_tooling_missing",
            "backend_preflight_failed",
            "circom_compilation_failed",
            "key_generation_failed",
            "missing_invariants",
            "readiness_failed",
            "filesystem_permission_denied",
        ],
    ),
    ("timeouts", &["wall_clock_timeout"]),
    (
        "stability_runtime",
        &[
            "runtime_error",
            "panic",
            "artifact_mirror_panic_missing_command",
            "run_outcome_missing",
            "run_outcome_unreadable",
            "run_outcome_invalid_json",
            "unknown",
            "stale_interrupted",
        ],
    ),
    (
        "contract_or_config",
        &["engagement_contract_failed", "missing_chains_definition"],
    ),
];

pub fn default_thresholds() -> BTreeMap<String, f64> {
    BTreeMap::from([
        ("lock_contention".to_string(), 0.05),
        ("setup_tooling".to_string(), 0.15),
        ("timeouts".to_string(), 0.10),
        ("stability_runtime".to_string(), 0.05),
        ("contract_or_config".to_string(), 0.10),
        ("other_failure".to_string(), 0.05),
    ])
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FailureClassRow {
    pub class: String,
    pub count: usize,
    pub rate: f64,
    pub max_rate: f64,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FailureDashboardPayload {
    pub generated_utc: Option<String>,
    pub summary_path: String,
    pub outcomes_path: String,
    pub total_runs: usize,
    pub overall_status: String,
    pub class_rows: Vec<FailureClassRow>,
    pub reason_counts: BTreeMap<String, usize>,
}

pub fn latest_benchmark_dir(benchmark_root: &Path) -> anyhow::Result<PathBuf> {
    let bench_name_re =
        Regex::new(r"^benchmark_[0-9]{8}_[0-9]{6}$").with_context(|| "Invalid benchmark regex")?;
    let mut candidates = Vec::new();
    for entry in std::fs::read_dir(benchmark_root)
        .with_context(|| format!("Failed to read '{}'", benchmark_root.display()))?
    {
        let entry = match entry {
            Ok(value) => value,
            Err(_) => continue,
        };
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        let Some(name) = path.file_name().and_then(|value| value.to_str()) else {
            continue;
        };
        if !bench_name_re.is_match(name) {
            continue;
        }
        if path.join("summary.json").is_file() && path.join("outcomes.json").is_file() {
            candidates.push(path);
        }
    }
    candidates.sort();
    candidates.into_iter().next_back().ok_or_else(|| {
        anyhow::anyhow!(
            "No benchmark_<timestamp> directory with summary/outcomes found under {}",
            benchmark_root.display()
        )
    })
}

pub fn load_json(path: &Path) -> anyhow::Result<Value> {
    let raw = std::fs::read_to_string(path)
        .with_context(|| format!("Failed to read '{}'", path.display()))?;
    serde_json::from_str(&raw).with_context(|| format!("Failed to parse '{}'", path.display()))
}

fn value_to_i64(value: &Value) -> Option<i64> {
    if let Some(number) = value.as_i64() {
        return Some(number);
    }
    if let Some(number) = value.as_u64() {
        return i64::try_from(number).ok();
    }
    if let Some(number) = value.as_f64() {
        return Some(number as i64);
    }
    value.as_str().and_then(|raw| raw.parse::<i64>().ok())
}

fn value_to_usize(value: &Value) -> Option<usize> {
    value_to_i64(value).and_then(|number| {
        if number < 0 {
            None
        } else {
            usize::try_from(number as u64).ok()
        }
    })
}

fn parse_rate(raw: &str, source: &str) -> anyhow::Result<f64> {
    let value: f64 = raw
        .parse()
        .with_context(|| format!("Invalid threshold value from {source}: {raw:?}"))?;
    if !(0.0..=1.0).contains(&value) {
        bail!("Threshold from {source} must be between 0.0 and 1.0, got {value}");
    }
    Ok(value)
}

pub fn env_key_for_class(class_name: &str) -> String {
    format!("ZKF_FAILURE_MAX_RATE_{}", class_name.to_ascii_uppercase())
}

pub fn resolve_thresholds(cli_overrides: &[String]) -> anyhow::Result<BTreeMap<String, f64>> {
    let mut thresholds = default_thresholds();

    for class_name in FAILURE_CLASSES {
        let env_key = env_key_for_class(class_name);
        let raw_env = match env::var(&env_key) {
            Ok(value) => value,
            Err(_) => continue,
        };
        if raw_env.trim().is_empty() {
            continue;
        }
        let parsed = parse_rate(raw_env.trim(), &format!("${env_key}"))?;
        thresholds.insert((*class_name).to_string(), parsed);
    }

    let known_classes: BTreeSet<String> = thresholds.keys().cloned().collect();
    for override_value in cli_overrides {
        let Some((raw_class, raw_rate)) = override_value.split_once('=') else {
            bail!("Invalid --threshold {override_value:?}; expected CLASS=RATE");
        };
        let class_name = raw_class.trim().to_ascii_lowercase();
        if !known_classes.contains(&class_name) {
            let valid = FAILURE_CLASSES.join(", ");
            bail!("Unknown failure class {class_name:?}; valid classes: {valid}");
        }
        let source = format!("--threshold {}={}", raw_class.trim(), raw_rate.trim());
        let parsed = parse_rate(raw_rate.trim(), &source)?;
        thresholds.insert(class_name, parsed);
    }
    Ok(thresholds)
}

pub fn classify_reason(reason: &str) -> &'static str {
    if SUCCESS_REASONS.contains(&reason) {
        return "success";
    }
    for (class_name, reasons) in FAILURE_CLASS_RULES {
        if reasons.contains(&reason) {
            return class_name;
        }
    }
    "other_failure"
}

pub fn sum_reason_counts(outcomes: &[Value]) -> BTreeMap<String, usize> {
    let mut counts: BTreeMap<String, usize> = BTreeMap::new();
    for item in outcomes {
        let Some(reason_counts) = item.get("reason_counts").and_then(Value::as_object) else {
            continue;
        };
        for (reason, raw_count) in reason_counts {
            let Some(parsed_count) = value_to_i64(raw_count) else {
                continue;
            };
            let non_negative = usize::try_from(parsed_count.max(0) as u64).unwrap_or(0);
            let entry = counts.entry(reason.clone()).or_insert(0);
            *entry = (*entry).saturating_add(non_negative);
        }
    }
    counts
}

pub fn dashboard(
    summary: &Value,
    outcomes: &[Value],
    summary_path: &Path,
    outcomes_path: &Path,
    class_thresholds: &BTreeMap<String, f64>,
) -> anyhow::Result<FailureDashboardPayload> {
    let generated_utc = summary
        .get("generated_utc")
        .and_then(Value::as_str)
        .map(str::to_string);

    let mut total_runs = summary
        .get("total_runs")
        .and_then(value_to_usize)
        .unwrap_or(outcomes.len());
    if total_runs == 0 {
        total_runs = outcomes.len();
    }

    let reason_counts = sum_reason_counts(outcomes);
    let mut class_counts: BTreeMap<String, usize> = BTreeMap::new();
    for (reason, count) in &reason_counts {
        let class_name = classify_reason(reason).to_string();
        let entry = class_counts.entry(class_name).or_insert(0);
        *entry = (*entry).saturating_add(*count);
    }

    let mut class_rows = Vec::new();
    for class_name in FAILURE_CLASSES {
        let count = class_counts.get(*class_name).copied().unwrap_or(0);
        let threshold = class_thresholds.get(*class_name).copied().ok_or_else(|| {
            anyhow::anyhow!("Missing threshold for failure class '{}'", class_name)
        })?;
        let rate = if total_runs > 0 {
            count as f64 / total_runs as f64
        } else {
            0.0
        };
        let status = if rate <= threshold { "PASS" } else { "FAIL" };
        class_rows.push(FailureClassRow {
            class: (*class_name).to_string(),
            count,
            rate,
            max_rate: threshold,
            status: status.to_string(),
        });
    }

    let overall_status = if class_rows.iter().any(|row| row.status == "FAIL") {
        "FAIL"
    } else {
        "PASS"
    };

    Ok(FailureDashboardPayload {
        generated_utc,
        summary_path: summary_path.display().to_string(),
        outcomes_path: outcomes_path.display().to_string(),
        total_runs,
        overall_status: overall_status.to_string(),
        class_rows,
        reason_counts,
    })
}

fn pct(value: f64) -> String {
    format!("{:.2}%", value * 100.0)
}

pub fn write_markdown(path: &Path, payload: &FailureDashboardPayload) -> anyhow::Result<()> {
    let mut lines = Vec::new();
    lines.push("# Benchmark Failure-Class Dashboard".to_string());
    lines.push(String::new());
    lines.push(format!(
        "- Generated UTC: `{}`",
        payload.generated_utc.as_deref().unwrap_or("n/a")
    ));
    lines.push(format!("- Summary: `{}`", payload.summary_path));
    lines.push(format!("- Outcomes: `{}`", payload.outcomes_path));
    lines.push(format!("- Overall status: `{}`", payload.overall_status));
    lines.push(format!("- Total runs: `{}`", payload.total_runs));
    lines.push(String::new());
    lines.push("| Failure Class | Count | Rate | Max Allowed | Status |".to_string());
    lines.push("|---|---:|---:|---:|---|".to_string());
    for row in &payload.class_rows {
        lines.push(format!(
            "| {} | {} | {} | {} | {} |",
            row.class,
            row.count,
            pct(row.rate),
            pct(row.max_rate),
            row.status
        ));
    }
    lines.push(String::new());
    lines.push("## Reason Counts".to_string());
    lines.push(String::new());
    lines.push("| Reason Code | Count |".to_string());
    lines.push("|---|---:|".to_string());
    for (reason, count) in &payload.reason_counts {
        lines.push(format!("| {} | {} |", reason, count));
    }
    lines.push(String::new());

    std::fs::write(path, lines.join("\n")).with_context(|| {
        format!(
            "Failed writing failure dashboard markdown '{}'",
            path.display()
        )
    })?;
    Ok(())
}
